File: Forms/frmProject.cs
using Microsoft.VisualBasic;
using System;
using System.Windows.Forms;

namespace eLib
    {
    public partial class frmProject
        {
        public frmProject ()
            {
            InitializeComponent ();
            }
        private void frmProject_Load (object sender, EventArgs e)
            {
            Width = 500;
            Height = 225;
            //Client.DialogRequestParams: bits: 1:Project 2:SubProjects 3:User Add/Edit
            //Client.DialogRequestParams bits: 4: 0:new 1:edit
            switch (Client.DialogRequestParams)
                {
                case 1: //project + new
                        {
                        Text = "NOTE: New Project:";
                        txtProjectName.Text = "Prj " + DateTime.Now.ToString ("yyyy.MM.dd-HH:mm") + "- EDIT";
                        txtProjectNote.Text = "note";
                        CheckBoxActive.Checked = true;
                        txtProjectName.Focus ();
                        txtProjectName.SelectionStart = 0;
                        txtProjectName.SelectionLength = Strings.Len (txtProjectName.Text);
                        break;
                        }
                case 9: //project + edit 
                        {
                        Text = "Edit Project:";
                        txtProjectName.Text = Project.Name;
                        txtProjectNote.Text = Project.Note;
                        CheckBoxActive.Checked = Project.IsActive;
                        txtProjectName.Focus ();
                        txtProjectName.SelectionStart = 0;
                        txtProjectName.SelectionLength = Strings.Len (txtProjectName.Text);
                        break;
                        }
                case 2: //subProject + new 
                        {
                        Text = "subProject:";
                        CheckBoxActive.Enabled = false;
                        Text = "New subProject:";
                        txtProjectName.Text = SubProject.Name;
                        txtProjectNote.Text = SubProject.Note;
                        txtProjectName.Focus ();
                        txtProjectName.SelectionStart = 0;
                        txtProjectName.SelectionLength = Strings.Len (txtProjectName.Text);
                        break;
                        }
                case 10: //subProject + edit 
                        {
                        Text = "subProject:";
                        CheckBoxActive.Enabled = false;
                        Text = "Edit subProject:";
                        txtProjectName.Text = Project.Name;
                        txtProjectNote.Text = Project.Note;
                        txtProjectName.Focus ();
                        txtProjectName.SelectionStart = 0;
                        txtProjectName.SelectionLength = Strings.Len (txtProjectName.Text);
                        break;
                        }
                case 4: //userPassword + new 
                        {
                        CheckBoxActive.Enabled = true;
                        Text = "New User:";
                        txtProjectName.Text = "User-" + DateTime.Now.ToString ("HHmmss") + "   - EDIT";
                        txtProjectNote.Text = "[password]";
                        // txtProjectNote.PasswordChar = "-"
                        txtProjectName.Focus ();
                        txtProjectName.SelectionStart = 0;
                        txtProjectName.SelectionLength = Strings.Len (txtProjectName.Text);
                        break;
                        }
                case 12: //userPassword + edit 
                        {
                        Text = "Change Password:";
                        txtProjectName.Text = Project.Name;
                        CheckBoxActive.Checked = Project.IsActive;
                        txtProjectName.Enabled = false;
                        txtProjectNote.Focus ();
                        txtProjectNote.Text = Project.Note;
                        txtProjectNote.SelectionStart = 0;
                        txtProjectNote.SelectionLength = Strings.Len (txtProjectNote.Text);
                        break;
                        }
                case 128: //new Student (bit8 on, bit4 off)
                        {
                        CheckBoxActive.Enabled = false;
                        Text = "New Student:";
                        txtProjectName.Text = Project.Name;
                        txtProjectNote.Text = Project.Note;
                        txtProjectName.Focus ();
                        txtProjectName.SelectionStart = 0;
                        txtProjectName.SelectionLength = Strings.Len (txtProjectName.Text);
                        break;
                        }
                case 136: //edit Student (bit8 on, bit4 on)
                        {
                        CheckBoxActive.Enabled = false;
                        Text = "Edit Student:";
                        txtProjectName.Text = Project.Name;
                        CheckBoxActive.Checked = true;
                        txtProjectName.Enabled = true;
                        txtProjectNote.Text = Project.Note;
                        txtProjectName.Focus ();
                        txtProjectName.SelectionStart = 0;
                        txtProjectName.SelectionLength = Strings.Len (txtProjectName.Text);
                        break;
                        }
                }
            }
        private void frmProject_KeyDown (object sender, KeyEventArgs e)
            {
            if (e.KeyCode.ToString () == "F5")
                {
                e.SuppressKeyPress = true;
                lblSave_Click (null, null);
                }
            else if (e.KeyCode.ToString () == "Escape")
                {
                e.SuppressKeyPress = true;
                lblCancel_Click (null, null);
                }
            }
        private void txtProjectName_KeyDown (object sender, KeyEventArgs e)
            {
            switch (e.KeyCode)
                {
                case Keys.Down:
                case Keys.Enter:
                        {
                        e.SuppressKeyPress = true;
                        txtProjectNote.Focus ();
                        txtProjectNote.SelectionStart = 0;
                        txtProjectNote.SelectionLength = Strings.Len (txtProjectNote.Text);
                        break;
                        }
                case Keys.Escape:
                        {
                        e.SuppressKeyPress = true;
                        Menu_Cancel_Click (null, null);
                        break;
                        }
                }
            }
        private void txtProjectNote_KeyDown (object sender, KeyEventArgs e)
            {
            switch (e.KeyCode)
                {
                case Keys.Enter:
                        {
                        if (Strings.Right (txtProjectName.Text, 6) == "- EDIT")
                            {
                            txtProjectName.Focus ();
                            txtProjectName.SelectionStart = 0;
                            txtProjectName.SelectionLength = Strings.Len (txtProjectName.Text);
                            e.SuppressKeyPress = true;
                            return;
                            }
                        if (Strings.Trim (txtProjectNote.Text) == "[password]")
                            {
                            txtProjectNote.SelectionStart = 0;
                            txtProjectNote.SelectionLength = Strings.Len (txtProjectNote.Text);
                            e.SuppressKeyPress = true;
                            return;
                            }
                        Menu_Save_Click (sender, e);
                        break;
                        }
                case Keys.Up:
                        {
                        try
                            {
                            e.SuppressKeyPress = true;
                            txtProjectName.Focus ();
                            }
                        catch { }
                        break;
                        }
                }
            }
        private void lblSave_Click (object sender, EventArgs e)
            {
            DoSave ();
            }
        private void Menu_Save_Click (object sender, EventArgs e)
            {
            DoSave ();
            }
        private void DoSave ()
            {
            if (Client.DialogRequestParams == 4)
                {
                //mode:userPassNew: chack space does not exist in username
                if ((string.IsNullOrEmpty (Strings.Trim (txtProjectName.Text))) || (txtProjectName.Text.Contains (" ")))
                    {
                    MessageBox.Show ("Notice: remove space(s) from Username");
                    txtProjectName.Focus ();
                    txtProjectName.SelectionStart = 0;
                    txtProjectName.SelectionLength = txtProjectName.Text.Length;
                    return;
                    }
                }
            Project.Name = txtProjectName.Text;
            Project.Note = txtProjectNote.Text;
            Project.IsActive = CheckBoxActive.Checked;
            if (string.IsNullOrEmpty (Strings.Trim (Project.Name)))
                {
                txtProjectName.Focus ();
                return;
                }
            else
                {
                Client.DialogRequestParams = 16; //set bit5 (00010000): 0:cancel, 1:save 
                Dispose ();
                }
            }
        private void lblCancel_Click (object sender, EventArgs e)
            {
            Menu_Cancel_Click (null, null);
            }
        private void Menu_Cancel_Click (object sender, EventArgs e)
            {
            Client.DialogRequestParams = 0; //set bit5 (00010000): 0:cancel, 1:save 
            Dispose ();
            }
        }
    }

File: Forms/frmProject.vb
Public Class frmProject
    Private Sub frmProject_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        'Retval1 {0:Project 1:subProject 2:User Add/Edit}
        'Retval2 {MODE 0:new 1:edit}
        'Retval3 {0:InActive -1:Active}

        Select Case Retval1
            Case 0 'Request from project
                Select Case Retval2 'MODE {new|edit}
                    Case 0 'new project
                        Me.Text = "NOTE: New Project:"
                        txtProjectName.Text = "Prj " & System.DateTime.Now.ToString("yyyy.MM.dd-HH:mm") & "- EDIT me!"
                        txtProjectNote.Text = "new project"
                        CheckBoxActive.Checked = True
                    Case 1 'edit mode
                        Me.Text = "Edit Project:"
                        txtProjectName.Text = strProjectName
                        txtProjectNote.Text = strProjectNote
                        CheckBoxActive.Checked = Retval3
                End Select
                txtProjectName.Focus()
                txtProjectName.SelectionStart = 0
                txtProjectName.SelectionLength = Len(txtProjectName.Text)
            Case 1 'Request from subProject
                Me.Text = "subProject:"
                CheckBoxActive.Enabled = False
                Select Case Retval2 'MODE {new|edit}
                    Case 0 'new subproject
                        Me.Text = "New subProject:"
                        txtProjectName.Text = "Prd " & System.DateTime.Now.ToString("yyyy.MM.dd-HH:mm") & "- EDIT me!"
                        txtProjectNote.Text = "[NOTE: new subproject"
                    Case 1 'edit mode
                        Me.Text = "Edit subProject:"
                        txtProjectName.Text = strProjectName
                        txtProjectNote.Text = strProjectNote
                End Select
                txtProjectName.Focus()
                txtProjectName.SelectionStart = 0
                txtProjectName.SelectionLength = Len(txtProjectName.Text)
            Case 2 'Request from USER Add/Edit
                '[password: type here]

                CheckBoxActive.Enabled = True
                Select Case Retval2
                    Case 0 'new User
                        Me.Text = "New User:"
                        txtProjectName.Text = "usr-" & System.DateTime.Now.ToString("MMddHHmm") & "- EDIT me!"
                        txtProjectNote.Text = "[password:type here]"
                        'txtProjectNote.PasswordChar = "-"
                        txtProjectName.Focus()
                        txtProjectName.SelectionStart = 0
                        txtProjectName.SelectionLength = Len(txtProjectName.Text)
                    Case 1
                        Me.Text = "Current Password:"
                        txtProjectName.Text = strProjectName
                        txtProjectName.Enabled = False
                        txtProjectNote.Text = ""
                        txtProjectNote.PasswordChar = "-"
                        CheckBoxActive.Checked = Retval3
                        txtProjectNote.Focus()
                    Case 2
                        Me.Text = "NEW Password:"
                        txtProjectName.Text = strProjectName
                        txtProjectName.Enabled = False
                        txtProjectNote.Text = ""
                        txtProjectNote.PasswordChar = "-"
                        CheckBoxActive.Checked = Retval3
                        txtProjectNote.Focus()
                    Case 3
                        Me.Text = "NEW Password (Confirm):"
                        txtProjectName.Text = strProjectName
                        txtProjectName.Enabled = False
                        txtProjectNote.Text = ""
                        txtProjectNote.PasswordChar = "-"
                        CheckBoxActive.Checked = Retval3
                        CheckBoxActive.Enabled = False
                        txtProjectNote.Focus()
                End Select
        End Select
    End Sub
    Private Sub txtProjectName_KeyDown(sender As Object, e As KeyEventArgs) Handles txtProjectName.KeyDown
        If e.KeyCode = 13 Then
            txtProjectNote.Focus()
            txtProjectNote.SelectionStart = 0
            txtProjectNote.SelectionLength = Len(txtProjectNote.Text)
            e.SuppressKeyPress = True
        End If
    End Sub
    Private Sub txtProjectNote_KeyDown(sender As Object, e As KeyEventArgs) Handles txtProjectNote.KeyDown
        If e.KeyCode = 13 Then
            If Microsoft.VisualBasic.Right(txtProjectName.Text, 10) = "- EDIT me!" Then
                txtProjectName.Focus()
                txtProjectName.SelectionStart = 0
                txtProjectName.SelectionLength = Len(txtProjectName.Text)
                e.SuppressKeyPress = True
                Exit Sub
            End If
            If Trim(txtProjectNote.Text) = "[password:type here]" Then
                txtProjectNote.SelectionStart = 0
                txtProjectNote.SelectionLength = Len(txtProjectNote.Text)
                e.SuppressKeyPress = True
                Exit Sub
            End If
            Menu_Save_Click(sender, e)
        End If
        '//Save and Exit
    End Sub
    Private Sub Menu_Save_Click(sender As Object, e As EventArgs) Handles Menu_Save.Click
        strProjectName = txtProjectName.Text
        strProjectNote = txtProjectNote.Text
        Retval3 = CheckBoxActive.Checked
        If Trim(strProjectName) = "" Then
            txtProjectName.Focus()
            Exit Sub
        Else
            Retval1 = 1 'send signal for a save (0: Cancel)
            Me.Dispose()
        End If

    End Sub
    Private Sub Menu_Cancel_Click(sender As Object, e As EventArgs) Handles Menu_Cancel.Click
        Retval1 = 0
        Me.Dispose()

    End Sub

End Class